File: virtual contest/contest1/findShortestSubArray/findShortestSubArray.py
class Solution:
    def findShortestSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        #length of seq of x == right(x) - left(x) + 1

        """
        object list in python3 : 
        create : A = {}
        set val: A[x] = ...
        iterate: for a in A: ...
        get values : A.values()
        """

        left,right,count = {},{},{}
        i = 0
        while i < len(nums):
            if nums[i] not in left:
                left[nums[i]] = i
            if nums[i] not in count:
                count[nums[i]] = 1
            elif nums[i] in count:
                count[nums[i]] += 1
            right[nums[i]] = i
            i += 1
            
        deg = max(count.values()) #degree
        length = 50000
        for n in count:
            if count[n] == deg:
                if length > right[n] - left[n] + 1:
                    length = right[n] - left[n] + 1
        
        return length
        
        

File: virtual contest/contest1/findShortestSubArray/findShortestSubArray.cpp
class Solution {
public:
    int findShortestSubArray(vector<int>& nums) {
        map<int,int> l;
        map<int,int> r;
        map<int,int> n;
        int max_num = 0;
        int min_len = 100000;
        for(int i = 0; i < nums.size(); i++){
            if(!l.count(nums[i]))
                l[nums[i]] = i;
            if(!n.count(nums[i]))
                n[nums[i]] = 1;
            else if(n.count(nums[i]))
                ++n[nums[i]];
        }
        
        for(int i = nums.size()-1; i >= 0; i--){
            max_num = max(max_num,n[nums[i]]);
            if(!r.count(nums[i]))
                r[nums[i]] = i;
        }
        
        for(int i = 0; i < nums.size(); i++){
            if(n[nums[i]] == max_num){
                min_len = min(min_len,r[nums[i]]-l[nums[i]]+1);
            }
        }
        
        return min_len;
    }
};